src/sparse/impl/KokkosSparse_bspgemm_impl_speed.hpp

```diff
@@ -0,0 +1,652 @@
+//
+// Questions? Contact Siva Rajamanickam (srajama@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include "KokkosKernels_Utils.hpp"
+
+namespace KokkosSparse {
+
+namespace Impl {
+
+template <typename HandleType, typename a_row_view_t_,
+          typename a_lno_nnz_view_t_, typename a_scalar_nnz_view_t_,
+          typename b_lno_row_view_t_, typename b_lno_nnz_view_t_,
+          typename b_scalar_nnz_view_t_>
+template <typename a_row_view_t, typename a_nnz_view_t,
+          typename a_scalar_view_t, typename b_row_view_t,
+          typename b_nnz_view_t, typename b_scalar_view_t,
+          typename c_row_view_t, typename c_nnz_view_t,
+          typename c_scalar_view_t, typename mpool_type>
+struct KokkosBSPGEMM<HandleType, a_row_view_t_, a_lno_nnz_view_t_,
+                     a_scalar_nnz_view_t_, b_lno_row_view_t_, b_lno_nnz_view_t_,
+                     b_scalar_nnz_view_t_>::NumericCMEM_CPU {
+  using BSPGEMM = KokkosBSPGEMM<HandleType, a_row_view_t_, a_lno_nnz_view_t_,
+                                a_scalar_nnz_view_t_, b_lno_row_view_t_,
+                                b_lno_nnz_view_t_, b_scalar_nnz_view_t_>;
+
+  nnz_lno_t numrows;
+  nnz_lno_t numcols;
+  nnz_lno_t block_dim;
+  nnz_lno_t block_size;
+
+  a_row_view_t row_mapA;
+  a_nnz_view_t entriesA;
+  a_scalar_view_t valuesA;
+
+  b_row_view_t row_mapB;
+  b_nnz_view_t entriesB;
+  b_scalar_view_t valuesB;
+
+  c_row_view_t rowmapC;
+  c_nnz_view_t entriesC;
+  c_scalar_view_t valuesC;
+  mpool_type memory_space;
+
+  nnz_lno_t *pEntriesC;
+  scalar_t *pVals;
+  const KokkosKernels::Impl::ExecSpaceType my_exec_space;
+  const nnz_lno_t team_work_size;
+
+  NumericCMEM_CPU(nnz_lno_t m_, nnz_lno_t k_, nnz_lno_t block_dim_,
+                  a_row_view_t row_mapA_, a_nnz_view_t entriesA_,
+                  a_scalar_view_t valuesA_,
+
+                  b_row_view_t row_mapB_, b_nnz_view_t entriesB_,
+                  b_scalar_view_t valuesB_,
+
+                  c_row_view_t rowmapC_, c_nnz_view_t entriesC_,
+                  c_scalar_view_t valuesC_, mpool_type memory_space_,
+                  const KokkosKernels::Impl::ExecSpaceType my_exec_space_,
+                  nnz_lno_t team_row_chunk_size)
+      : numrows(m_),
+        numcols(k_),
+        block_dim(block_dim_),
+        block_size(block_dim_ * block_dim_),
+        row_mapA(row_mapA_),
+        entriesA(entriesA_),
+        valuesA(valuesA_),
+
+        row_mapB(row_mapB_),
+        entriesB(entriesB_),
+        valuesB(valuesB_),
+
+        rowmapC(rowmapC_),
+        entriesC(entriesC_),
+        valuesC(valuesC_),
+        memory_space(memory_space_),
+        pEntriesC(entriesC_.data()),
+        pVals(valuesC.data()),
+        my_exec_space(my_exec_space_),
+        team_work_size(team_row_chunk_size) {}
+
+  KOKKOS_INLINE_FUNCTION
+  size_t get_thread_id(const size_t row_index) const {
```

Comment:
It's not mine and I haven't changed it, but here's how I read the semantics:  
Essentially we use `get_thread_id()` to _index thread's memory chunk_ with `allocate_chunk()` on `UniformMemoryPool` (defined [here](https://github.com/kokkos/kokkos-kernels/blob/403a4177774fda29426e289367de705bf9cbe0a8/src/sparse/impl/KokkosSparse_spgemm_impl_kkmem.hpp#L1596-L1604)):
* on GPU the pool is `ManyThread2OneChunk` and chunk index is effectively `row_index % (concurrency / suggested_vector_size)` (`row_index` just goes through `get_thread_id()` which doesn't calculate anything in this case).
* on CPU the pool is `OneThread2OneChunk` and chunk index is:
  -  `0` for Serial;
  - `thread_id % num_chunks` for OpenMP/Threads where `thread_id` is actual thread id obtained with `get_thread_id()` (which ignores input `row_index` in this case...) and `num_chunks` is calculated with [`compute_num_pool_chunks()`](https://github.com/kokkos/kokkos-kernels/blob/403a4177774fda29426e289367de705bf9cbe0a8/src/sparse/impl/KokkosSparse_spgemm_impl.hpp#L807-L840).

PS. I hope `get_thread_id()` can be refactored at some point because:
1. naming and semantics is quite unclear,
2. it's duplicated few times in spgemm files (e.g. see out of sync variants in [`sparseacc`](https://github.com/kokkos/kokkos-kernels/blob/403a4177774fda29426e289367de705bf9cbe0a8/src/sparse/impl/KokkosSparse_spgemm_jacobi_sparseacc_impl.hpp#L249-L271) and [`denseacc`](https://github.com/kokkos/kokkos-kernels/blob/403a4177774fda29426e289367de705bf9cbe0a8/src/sparse/impl/KokkosSparse_spgemm_jacobi_denseacc_impl.hpp#L116-L132) jacobi),
3. gets short-circuited [here](https://github.com/kokkos/kokkos-kernels/blob/403a4177774fda29426e289367de705bf9cbe0a8/src/sparse/impl/KokkosSparse_spgemm_impl_kkmem.hpp#L554-L558) and [here](https://github.com/kokkos/kokkos-kernels/blob/403a4177774fda29426e289367de705bf9cbe0a8/src/sparse/impl/KokkosSparse_spgemm_impl_kkmem.hpp#L710-L711)...